probe-rs/src/architecture/arm/sequences/infineon.rs

```diff
@@ -0,0 +1,249 @@
+//! Sequences for Infineon target families
+
+use crate::architecture::arm::armv7m::{Aircr, Dhcsr, FpCtrl, FpRev1CompX, FpRev2CompX};
+use anyhow::anyhow;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::time::{Duration, Instant};
+
+use crate::Error;
+use crate::Memory;
+use crate::{DebugProbeError, MemoryMappedRegister};
+
+use super::ArmDebugSequence;
+
+/// An Infineon XMC4xxx MCU.
+pub struct XMC4000 {
+    halt_after_reset_state: Mutex<Option<(bool, u32)>>,
```

Comment:
Silly question: why is this in a Mutex? Can't we do it without inner mutability?

Reply:
`reset_catch_set()` sets a breakpoint and `reset_catch_clear()` clears it. There's a fixed number of hardware breakpoints, and using one to halt after reset clobbers one. Resetting doesn't normally disrupt debugging hardware, and the `ResetCatchSet` sequence [specifically describes](https://open-cmsis-pack.github.io/Open-CMSIS-Pack-Spec/main/html/debug_description.html#default_sequences) `ResetCatchClear` cleaning up afterwards:

> The implementation of ResetCatchSet requires an implementation of ResetCatchClear to free any hardware resources used for stopping the core.

I took this to mean that `reset_catch_clear()` should restore the breakpoint, which means `reset_catch_set()` needs to squirrel it away: https://github.com/probe-rs/probe-rs/blob/eb60a3522e9cc340a7174e324ff92f9d0695ba24/probe-rs/src/architecture/arm/sequences/infineon.rs#L46-L48

Both take `&self`, so I used a `Mutex` to store it inside `struct XMC4000`.

Reply:
I gave this tuple a name in https://github.com/probe-rs/probe-rs/pull/1301/commits/8d216ff3f2cefb723ce9cd660bbcbde607923d4c.